integration_tests/steps/infinite_bus_tests.py

```diff
@@ -37,7 +37,7 @@ def check_buy_behaviour_ib(context):
         for trade in core_stats[name_uuid_map['Grid']]['trades']:
             assert limit_float_precision(trade['energy_rate']) <= \
                    core_stats[name_uuid_map['Infinite Bus']]['energy_rate'] or \
-                   "Infinite Bus" is not trade['seller']
+                   "Infinite Bus" is not trade['seller']  # noqa: F632
```

Comment:
What is Flake complaining about?

Reply:
Thanks @hannesdiedrich, good spot. 
> F632 use ==/!= to compare str, bytes, and int literals

In theory, it should be working by replacing `is not` with `!=`. In practice, it does not.

```
Then Infinite Bus buys energy that is not needed from the PV and sells to the load                      # integration_tests/steps/infinite_bus_tests.py:30 0.006s
    Traceback (most recent call last):
      File "python3.8/site-packages/behave/model.py", line 1329, in run
        match.run(runner.context)
      File "python3.8/site-packages/behave/matchers.py", line 98, in run
        self.func(context, *args, **kwargs)
      File "integration_tests/steps/infinite_bus_tests.py", line 59, in check_buy_behaviour_ib
        assert limit_float_precision(trade['energy_rate']) <= \
    AssertionError
```

If I log the content of the variables for those lines, I get this:
```
limit_float_precision(trade['energy_rate']) ->  26.0
core_stats[name_uuid_map['Infinite Bus']]['energy_rate'] 24

# This should return True, but it's returning False
limit_float_precision(trade['energy_rate']) <= core_stats[name_uuid_map['Infinite Bus']]['energy_rate'] or ""Infinite Bus" != trade['seller'] ->  False

trade['seller'] ->  Infinite Bus
"Infinite Bus" != trade['seller'] ->  False
"Infinite Bus" is not trade['seller'] ->  True
```

The check on the strings returns a different result than before (when it was using `is not`). I this might be exposing some pre-existing problem somewhere else. What do you think?

Reply:
This is a real bug, and flake is correct to point it out actually. When comparing strings, we should almost always use `!=` and not the `is not`. I would say to remove the NOQA directive and use the inequality operator even now, to both satisfy Flake and solve the issue

Reply:
@spyrostz substituting `is not` with `!=` causes the tests to fail:
```
Scenario Outline: Infinite Bus behaviour -- @1.2 Market Type                                              # integration_tests/infinite_bus.feature:13
  Given we have a scenario named two_sided_market.infinite_bus                                            # integration_tests/steps/integration_tests.py:47 0.000s
  And d3a is installed                                                                                    # integration_tests/steps/integration_tests.py:56 0.000s
  And the market type is 2                                                                                # integration_tests/steps/infinite_bus_tests.py:25 0.000s
  And export is_needed                                                                                    # integration_tests/steps/integration_tests.py:639 0.000s
  When we run the simulation with setup file two_sided_market.infinite_bus and parameters [24, 60, 60, 1] # integration_tests/steps/integration_tests.py:645 1.595s
  Then Infinite Bus buys energy that is not needed from the PV and sells to the load                      # integration_tests/steps/infinite_bus_tests.py:30 0.007s
    Traceback (most recent call last):
      File "d3a/lib/python3.8/site-packages/behave/model.py", line 1329, in run
        match.run(runner.context)
      File "d3a/lib/python3.8/site-packages/behave/matchers.py", line 98, in run
        self.func(context, *args, **kwargs)
      File "integration_tests/steps/infinite_bus_tests.py", line 38, in check_buy_behaviour_ib
        assert limit_f  loat_precision(trade['energy_rate']) <= \
    AssertionError  
```

That was the reason why, at first, I thought the `is not` was meant to stay. I might need some more background knowledge about this test. I'll also test if replacing `is not` with `!=` works on the current master version with python 3.6 (I suppose it doesn't).

Reply:
I can confirm that replacing `is not` with `!=` generate failures in tests even on the current `master` branch with python 3.6.

Reply:
Tests fail on this trade:
```python
{
  'type': 'Trade',
  'match_type': 'Offer',
  'id': '978b2a75-b9aa-4972-a0fd-34c4d80eaa87',
  'offer_bid_id': '6f3cfba7-7553-4c76-881e-c4c52bcec76a',
  'residual_id': None,
  'energy': 0.2,
  'energy_rate': 26.0,
  'price': 5.2,
  'buyer': 'IAA House 1',
  'buyer_origin': 'H1 General Load',
  'seller_origin': 'Infinite Bus',
  'seller': 'Infinite Bus',
  'fee_price': 0.0,
  'time': '2020-12-16T07:00:00'
}
```
- `trade['energy_rate']`: `26.0`
- `core_stats[name_uuid_map['Infinite Bus']]['energy_rate']`: `24`
- `trade['seller']`: "Infinite Bus"

So the assertion is rightfully failing in this case.

Reply:
Please replace the assert with these checks:

@then('Infinite Bus buys energy that is not needed from the PV and sells to the load')
def check_buy_behaviour_ib(context):
    from integration_tests.steps.two_sided_market import no_unmatched_loads
    no_unmatched_loads(context)
    name_uuid_map = get_area_name_uuid_mapping(context.area_tree_summary_data)
    grid = context.simulation.area
    bus = list(filter(lambda x: x.name == "Infinite Bus", grid.children))[0]

    for time_slot, core_stats in context.raw_sim_data.items():
        for trade in core_stats[name_uuid_map['Grid']]['trades']:
            if trade['seller'] == "Infinite Bus":
                assert limit_float_precision(trade['energy_rate']) >= \
                       core_stats[name_uuid_map['Infinite Bus']]['energy_rate']
            else:
                assert limit_float_precision(trade['energy_rate']) <= \
                       core_stats[name_uuid_map['Infinite Bus']]['energy_rate']

            if trade['buyer'] == "Infinite Bus":
                # TODO: energy_buy_rate is not part of the raw simulation data, therefore
                # it is obligatory to retrieve it from the strategy. This needs to change once
                # we add the energy_buy_rate to the infinite bus result data.
                assert limit_float_precision(trade['energy_rate']) <= \
                       list(bus.strategy.energy_buy_rate.values())[0]

            if trade['buyer'] == "H1 General Load":
                assert trade['energy'] == \
                       core_stats[name_uuid_map['H1 General Load']]['load_profile_kWh']
            elif trade['seller'] == 'Infinite Bus':
                assert isclose(trade['energy'],
                               (core_stats[name_uuid_map['H1 General Load']]['load_profile_kWh'] -
                                core_stats[name_uuid_map['H1 PV']]['pv_production_kWh']))


The check is incorrect in the first place, because the condition `trade["energy_rate"] (26.0) <= infinite_bus_energy_rate (24.0)` cannot be true if the infinite bus is the seller (like the original assert implies). I tried to replace this check with an easier to read IMO if block, that states that if the infinite bus is the seller, then the trade should have a rate greater than the sell energy rate of the infinite bus (essentially, that the infinite bus sells with a minimum price of 24 cents). If the seller is not the infinite bus, then it is implied that a cheaper-than-the-infinite-bus producer was used for this trade (hence the trade rate should be less than or equal than the infinite bus). 

Also added a similar check for the buy behavior of the infinite bus since I was in the territory. 

Pinging @faizan2590 too in order to be aware of it.